design/checkpoint_03/harmony_py.ipynb

```diff
@@ -0,0 +1,286 @@
+{
+ "metadata": {
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.8.5-final"
+  },
+  "orig_nbformat": 2,
+  "kernelspec": {
+   "name": "python3",
+   "display_name": "Python 3.8.5 64-bit",
+   "metadata": {
+    "interpreter": {
+     "hash": "31862cc71836a94b0e0781803a3648767fc4cb197cc35bade0ddf231ddce7d7c"
+    }
+   }
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2,
+ "cells": [
+  {
+   "source": [
+    "# Usage Overview\n",
+    "The following is example usage for `harmony-py`. \n"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "source": [
+    "\n",
+    "# Under this proposed interface harmony-py would have the concept of three (3) entities: a Request, a Client, and a Job:\n",
+    "# - The Request contains request parameters\n",
+    "# - The Client represents the ability to perform authenticated HTTP with the Harmony endpoint\n",
+    "# - The Job represents what Harmony is working on as well as retrieving the finished work results. It's referenced via 'job_id' and used by the Client.\n",
+    "\n",
+    "# Individual parameters are validated when set.\n",
+    "# Not all keyword args need be supplied at once. Also, parameters may be replaced.\n",
+    "request = Request(\n",
+    "    collection=Collection(id='C1940468263-POCLOUD'),\n",
+    "    spatial={'ll': (40, -107),\n",
+    "             'ur': (42, -105)}\n",
+    "    temporal={'start': datetime.date(2020, 6, 1),\n",
+    "              'stop': datetime.date(2020, 6, 30)},\n",
+    "    format=Format.ZARR\n",
+    ")\n",
+    "\n",
+    "# Authentication is stored in a client object for subsequent server interaction.\n",
+    "client = Client(Authentication())\n",
+    "\n",
+    "# Validation may be performed prior to job processing; uses Harmony server-side checking.\n",
+    "client.validate(request)\n",
+    "\n",
+    "# Starts job processing; async by default.\n",
+    "job_id = client.submit(request, verbose=True, async=True)\n",
+    "\n",
+    "# Optional\n",
+    "client.status(job_id)\n",
+    "\n",
+    "# Optional\n",
+    "client.cancel(job_id)\n",
+    "\n",
+    "# Retrieve results in-region; returns a generator\n",
+    "urls = client.result_urls(job_id, region=Region.US_WEST_2)\n",
+    "# - or -\n",
+    "# Download files to a local directory\n",
+    "client.download(job_id, region=Region.US_WEST_2, directory='./research', overwrite=True)\n"
```

Comment:
How does `download` work when requesting data in region? Or should this only be an option if, say `region` isn't specified in this example, implying that the user wants the https urls?

Reply:
I had assumed people would want to download data no matter their location but if the library is being executed in region, should we only return a list of URLs?

Reply:
Yeah this is a good point. I have been thinking about these two binary groups of users: those who are fully outside of the cloud world, and those that only want to work with data in place within the cloud. I suppose there is nothing stopping a user from being able to download data directly from the Harmony API now even if they're running in-region, so this is valid. I don't think we want to be encouraging it though (or I shuold say, we should be encouraging analysis-in-place instead), so maybe that's where Im hung up. Worth keeping this here now so we can come back to it in the development story.